File: 프로그래머스/2/43165. 타겟 넘버/타겟 넘버.py
from collections import deque

def solution(numbers, target):
    answer = 0
    
    #bfs를 이용해줄거임, 현재 합이랑, 현재 인덱스를 넣어줌
    que=deque()
    que.append((0,0))
    while que:
        current_sum, index = que.popleft()
        if index==len(numbers):
            for a,b in que:
                if a==target:
                    answer+=1
            print(answer)
            break
        que.append((current_sum+numbers[index], index+1))
        que.append((current_sum-numbers[index], index+1))


    return answer

File: 프로그래머스/2/43165. 타겟 넘버/타겟 넘버.java
import java.util.*;
class Solution {
    static int nums[];
    static int target;
    static int answer;
    static int n;
    public int solution(int[] numbers, int targets) {
        answer = 0;
        nums = numbers;
        target = targets;
        n = numbers.length;
        dfs(0,0,0);
        return answer;
    }
    static void dfs(int cnt, int sum, int start) {
        if(cnt == n) {
            if(sum == target) answer++; //같으면
            return; 
        }
        for(int i=start; i<n; i++) {
            dfs(cnt+1, sum+nums[i], i+1);
            dfs(cnt+1, sum-nums[i], i+1);
        }
    }
}